pkg/controller/extensions/extension/extension.go

```diff
@@ -0,0 +1,251 @@
+package extension
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	controllerHandler "sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	corev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/apis/extensions/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/logging"
+)
+
+const (
+	controllerName = "extension.extensions.crossplane.io"
+
+	reconcileTimeout      = 1 * time.Minute
+	requeueAfterOnSuccess = 10 * time.Second
+
+	reasonCreatingRBAC       = "failed to create RBAC"
+	reasonCreatingDeployment = "failed to create deployment"
+)
+
+var (
+	log              = logging.Logger.WithName(controllerName)
+	resultRequeue    = reconcile.Result{Requeue: true}
+	requeueOnSuccess = reconcile.Result{RequeueAfter: requeueAfterOnSuccess}
+)
+
+// Reconciler reconciles a Instance object
+type Reconciler struct {
+	kube client.Client
+	factory
+}
+
+// Add creates a new Extension Controller and adds it to the Manager with default RBAC.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	r := &Reconciler{
+		kube:    mgr.GetClient(),
+		factory: &extensionHandlerFactory{},
+	}
+
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Extension
+	err = c.Watch(&source.Kind{Type: &v1alpha1.Extension{}}, &controllerHandler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile reads that state of the Extension for a Instance object and makes changes based on the state read
+// and what is in the Instance.Spec
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("reconciling", "kind", v1alpha1.ExtensionKindAPIVersion, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	// fetch the CRD instance
+	i := &v1alpha1.Extension{}
+	if err := r.kube.Get(ctx, req.NamespacedName, i); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	handler := r.factory.newHandler(ctx, i, r.kube)
+
+	return handler.sync(ctx)
+}
+
+type handler interface {
+	sync(context.Context) (reconcile.Result, error)
+	create(context.Context) (reconcile.Result, error)
+	update(context.Context) (reconcile.Result, error)
+}
+
+type extensionHandler struct {
+	kube client.Client
+	i    *v1alpha1.Extension
```

Comment:
Nit: I am :100:  in support of succinct (local) variable names, however, in the case of the struct field I think I prefer something more informative. 

```suggestion
	extension    *v1alpha1.Extension
```
or very least
```suggestion
	ext    *v1alpha1.Extension
```

Reply:
i'll make this change, the `i` was feeling too terse to me too